### PR TITLE
chore(invdes): fixed smoothed projection gradient with beta=inf

### PR DESCRIPTION
**Pull request by mahlau-flex**
Fixed the gradient computation in smoothed_projection for beta=inf. The problem was that the existing tanh_projection did not handle beta=inf properly.

<h2>Greptile Overview</h2>

<h3>Greptile Summary</h3>

Fixed gradient computation in `tanh_projection` when `beta=inf` by adding an explicit step function case using `np.where`, preventing NaN/inf gradients that occurred with the previous tanh formula. The fix enables proper gradient-based optimization in topology design with completely binarized projections.

Key changes:
- Added `beta == np.inf` special case in `tanh_projection` (projections.py:70-71) returning step function `np.where(array > eta, 1.0, 0.0)`
- Refactored tests to use `circle_2d_factory` pytest fixture eliminating code duplication
- Added comprehensive parametrized test `test_projection_gradient_correctness` using autograd's `check_grads` to verify forward and reverse mode gradients up to 2nd order
- Test coverage now includes `beta=np.inf` case across multiple array sizes, radii, and smoothing parameters

<h3>Confidence Score: 4/5</h3>

- This PR is safe to merge with minor considerations about edge case documentation
- The fix correctly addresses the gradient computation issue for `beta=inf` by using a step function instead of the tanh formula that causes numerical instability. Comprehensive test coverage validates gradients across multiple scenarios including the `beta=inf` case. The only minor concern is documentation clarity around the `array == eta` edge case behavior.
- No files require special attention - the implementation is sound and well-tested

<h3>Important Files Changed</h3>

File Analysis

| Filename | Score | Overview |
|----------|-------|----------|
| tidy3d/plugins/autograd/invdes/projections.py | 4/5 | Added special case handling for `beta=np.inf` in `tanh_projection` to prevent NaN/inf gradients by using step function |
| tests/test_plugins/autograd/invdes/test_projections.py | 5/5 | Refactored tests to use fixture for circle generation, added comprehensive gradient correctness tests with parametrized `beta=np.inf` cases |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant smoothed_projection
    participant tanh_projection
    participant autograd
    
    User->>smoothed_projection: call with array, beta=inf, eta
    smoothed_projection->>tanh_projection: compute original_projected
    
    alt beta == 0
        tanh_projection-->>smoothed_projection: return array (unchanged)
    else beta == inf (NEW FIX)
        tanh_projection->>tanh_projection: np.where(array > eta, 1.0, 0.0)
        tanh_projection-->>smoothed_projection: return step function
    else finite beta
        tanh_projection->>tanh_projection: compute tanh formula
        tanh_projection-->>smoothed_projection: return smooth projection
    end
    
    smoothed_projection->>smoothed_projection: compute gradients & smoothing
    smoothed_projection->>tanh_projection: compute rho_minus_eff_projected
    smoothed_projection->>tanh_projection: compute rho_plus_eff_projected
    smoothed_projection->>smoothed_projection: blend with polynom weights
    smoothed_projection-->>User: return smoothed result
    
    User->>autograd: compute gradient
    autograd->>smoothed_projection: backpropagate
    smoothed_projection->>tanh_projection: gradient through step function (when beta=inf)
    Note over tanh_projection,autograd: np.where has zero gradient<br/>(doesn't cause NaN/inf)
    autograd-->>User: return finite gradients
```

### CODE REVIEW

File: tests/test_plugins/autograd/invdes/test_projections.py [hunk 1]
```diff
@@ -2,23 +2,38 @@
...
 
 from tidy3d.plugins.autograd.invdes.filters import ConicFilter
 from tidy3d.plugins.autograd.invdes.projections import smoothed_projection, tanh_projection
 
 
-def test_smoothed_projection_beta_inf():
-    nx, ny = 50, 50
-    arr = np.zeros((50, 50), dtype=float)
+@pytest.fixture
+def circle_2d_factory():
```
Comment (yaugenst-flex) [inline tests/test_plugins/autograd/invdes/test_projections.py]: Wondering why we'd want this to be a fixture over just a helper function?

**Review by yaugenst-flex — APPROVED** [review]
LGTM!

File: tests/test_plugins/autograd/invdes/test_projections.py (hunk 1 above)
Comment (mahlau-flex) [reply tests/test_plugins/autograd/invdes/test_projections.py]: yes, this does not really make sense here since it is just a helper function. I have removed it
